File: scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/algorithms/PSI.scala
/*
 * Copyright 2021 The BigDL Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.bigdl.ppml.algorithms

import java.util
import java.util.concurrent.TimeoutException

import com.intel.analytics.bigdl.ppml.FLContext
import com.intel.analytics.bigdl.ppml.psi.HashingUtils
import com.intel.analytics.bigdl.ppml.utils.FLClientClosable
import org.apache.logging.log4j.LogManager
import org.apache.spark.sql.DataFrame

import collection.JavaConverters._
import collection.JavaConversions._
import scala.util.control.Breaks._

class PSI() extends FLClientClosable {
  private var hashIdToId: Map[String, String] = Map[String, String]()
  val logger = LogManager.getLogger(getClass)

  def getSalt(secureCode: String = ""): String = {
    flClient.psiStub.getSalt(secureCode)
  }

  def uploadSet(ids: util.List[String], salt: String): Unit = {
    val hashedIdArray = HashingUtils.parallelToSHAHexString(ids, salt)
    hashIdToId = hashedIdArray.zip(ids).toMap
    flClient.psiStub.uploadSet(hashedIdArray)
  }


  def downloadIntersection(maxTry: Int = 100, retry: Long = 3000): util.List[String] = {
    var intersection: util.List[String] = null
    breakable {
      for (i <- 0 until maxTry) {
        intersection = flClient.psiStub.downloadIntersection
        if (intersection == null) {
          if (i == maxTry - 1) {
            throw new TimeoutException("Max retry reached, could not get intersection, exited.")
          }
          logger.info(s"Got empty intersection, retry in $retry ms")
          Thread.sleep(retry)
        }
        else {
          logger.info("Intersection successful. Intersection's size is " + intersection.size + ".")
          break
        }
      }
    }
    intersection
  }
  def getIntersectionSet(df: DataFrame,
                         rowKeyName: String,
                         intersection: util.List[String]): DataFrame = {
    val intersectionSet = intersection.toSet
    val dataSet = df.filter(r => intersectionSet.contains(r.getAs[String](rowKeyName)))
    dataSet
  }
  def uploadSetAndDownloadIntersection(df: DataFrame,
                                       salt: String,
                                       rowKeyName: String = "ID",
                                       maxTry: Int = 100,
                                       retry: Long = 3000): DataFrame = {
    val spark = FLContext.getSparkSession()
    import spark.implicits._
    val ids = df.select(rowKeyName).as[String].collect().toList
    uploadSet(ids, salt)
    val hashIntersection = downloadIntersection(maxTry, retry)
    val intersection = hashIntersection.map(h => hashIdToId(h))
    getIntersectionSet(df, rowKeyName, intersection)
  }
}

File: python/ppml/src/bigdl/ppml/fl/algorithms/psi.py
#
# Copyright 2016 The BigDL Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#


from bigdl.ppml.fl import *

from bigdl.dllib.utils.common import JavaValue


def set_psi_salt(psi_salt):
    callBigDlFunc("float", "setPsiSalt", psi_salt)


class PSI(JavaValue):
    def __init__(self, jvalue=None, *args):
        self.bigdl_type = "float"
        super().__init__(jvalue, self.bigdl_type, *args)

    def get_salt(self, secure_code=""):
        return callBigDlFunc(self.bigdl_type, "psiGetSalt", self.value, secure_code)

    def upload_set(self, ids, salt):
        callBigDlFunc(self.bigdl_type, "psiUploadSet", self.value, ids, salt)

    def download_intersection(self, max_try=100, retry=3000):
        return callBigDlFunc(self.bigdl_type, "psiDownloadIntersection", self.value, max_try, retry)

    def get_intersection(self, ids, max_try=100, retry=3000):
        return callBigDlFunc(self.bigdl_type, "psiGetIntersection", \
            self.value, ids, max_try, retry)
